File: Assignment 4/doctor-soa/src/main/java/ds/health/endpoint/PatientRecommendationEndpoint.java
package ds.health.endpoint;

import ds.health.model.PatientEntity;
import ds.health.repository.PatientRepository;
import ds.health.soa.soa_endpoints.SetPatientRecommendationRequest;
import ds.health.soa.soa_endpoints.SetPatientRecommendationResponse;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.ws.server.endpoint.annotation.Endpoint;
import org.springframework.ws.server.endpoint.annotation.PayloadRoot;
import org.springframework.ws.server.endpoint.annotation.RequestPayload;
import org.springframework.ws.server.endpoint.annotation.ResponsePayload;

import java.util.Optional;

@Endpoint
public class PatientRecommendationEndpoint {
    private static final String NAMESPACE_URI = "http://health.ds/soa/soa-endpoints";

    private PatientRepository patientRepository;

    @Autowired
    public PatientRecommendationEndpoint(PatientRepository patientRepository) {
        this.patientRepository = patientRepository;
    }

    @PayloadRoot(namespace = NAMESPACE_URI, localPart = "setPatientRecommendationRequest")
    @ResponsePayload
    public SetPatientRecommendationResponse setPatientRecommendation(@RequestPayload SetPatientRecommendationRequest request) {

        SetPatientRecommendationResponse response = new SetPatientRecommendationResponse();

        Optional<PatientEntity> patientEntity = patientRepository.findById(request.getId());
        if (patientEntity.isPresent()) {
            patientEntity.get().setRecommendation(request.getRecommendation());
            patientRepository.save(patientEntity.get());
            response.setResponse("Patient recommendation set");
        } else {
            response.setResponse("Patient does not exist!");
        }

        return response;
    }
}

File: Assignment 4/caregiver-soa/src/main/kotlin/ds/health/caregiversoa/endpoint/PatientRecommendationEndpoint.kt
package ds.health.caregiversoa.endpoint

import ds.health.caregiversoa.repository.PatientRepository
import ds.health.soa.soa_endpoints.GetPatientRecommendationRequest
import ds.health.soa.soa_endpoints.GetPatientRecommendationResponse
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.ws.server.endpoint.annotation.Endpoint
import org.springframework.ws.server.endpoint.annotation.PayloadRoot
import org.springframework.ws.server.endpoint.annotation.RequestPayload
import org.springframework.ws.server.endpoint.annotation.ResponsePayload

@Endpoint
class PatientRecommendationEndpoint {

    @Autowired
    private lateinit var patientRepository: PatientRepository

    @PayloadRoot(namespace = "http://health.ds/soa/soa-endpoints", localPart = "getPatientRecommendationRequest")
    @ResponsePayload
    fun getPatientRecommendation(@RequestPayload request: GetPatientRecommendationRequest): GetPatientRecommendationResponse {

        val response = GetPatientRecommendationResponse()

        val patientEntity = patientRepository.findById(request.id)
        if (patientEntity.isPresent) {
            if (patientEntity.get().recommendation == null) response.recommendation = "" else response.recommendation = patientEntity.get().recommendation
        } else {
            response.recommendation = ""
        }

        return response
    }
}
